src/Entity/Citation.php

```diff
@@ -0,0 +1,377 @@
+
+    $fields['parent_id'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Parent ID'))
+      ->setDescription(t('The ID of the parent entity of which this entity is referenced.'))
+      ->setSetting('is_ascii', TRUE);
+
+    $fields['parent_type'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Parent type'))
+      ->setDescription(t('The entity parent type to which this entity is referenced.'))
+      ->setSetting('is_ascii', TRUE)
+      ->setSetting('max_length', EntityTypeInterface::ID_MAX_LENGTH);
+
+    $fields['parent_field_name'] = BaseFieldDefinition::create('string')
+      ->setLabel(t('Parent field name'))
+      ->setDescription(t('The entity parent field name to which this entity is referenced.'))
+      ->setSetting('is_ascii', TRUE)
+      ->setSetting('max_length', FieldStorageConfig::NAME_MAX_LENGTH);
+
+    return $fields;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getParentEntity() {
+    if (!isset($this->get('parent_type')->value) || !isset($this->get('parent_id')->value)) {
+      return NULL;
+    }
+
+    $parent = $this->entityTypeManager()
+      ->getStorage($this->get('parent_type')->value)
+      ->load($this->get('parent_id')->value);
+
+    // Return current translation of parent entity, if it exists.
+    if (
+      $parent != NULL &&
+      ($parent instanceof TranslatableInterface) &&
+      $parent->hasTranslation($this->language()->getId())
+    ) {
+      return $parent->getTranslation($this->language()->getId());
+    }
+
+    return $parent;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setParentEntity(ContentEntityInterface $parent, $parent_field_name) {
+    $this->set('parent_type', $parent->getEntityTypeId());
+    $this->set('parent_id', $parent->id());
+    $this->set('parent_field_name', $parent_field_name);
+    return $this;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getBibliography($style = self::APA): string {
+
+    $data = [
+      'id' => $this->id(),
+      'title' => $this->label(),
+      // Custom variables that wrap the title with a link tag.
+      'link-beginning' => $this->getLinkBeginning(),
+      'link-ending' => $this->getLinkBeginning() ? '</a>' : NULL,
+      'DOI' => $this->getDoi(),
+      'URL' => $this->getUrl(),
+      'author' => $this->getAuthor(),
+      'edition' => (int) $this->getEdition(),
+      'issue' => $this->getIssue(),
+      'issued' => $this->getDate(),
+      'genre' => $this->getGenre(),
+      'page' => $this->getPage(),
+      'publisher' => $this->getPublisher(),
+      'publisher-place' => $this->getPublisherPlace(),
+      'subtitle' => $this->getSubtitle(),
+      'type' => $this->getType(),
+      'volume' => (int) $this->getVolume(),
+    ];
+
+    if ($data['type'] == 'article-journal') {
+      $data['collection-title'] = $data['publisher'];
+    }
+
+    // Convert the arrays into objects.
+    $data = json_decode(json_encode([array_filter($data)]));
```

Comment:
Just thought I'd look up this approach and what benefits/losses it has from simple type casting. Looks like your gain of recursion is needed here

https://stackoverflow.com/questions/1869091/how-to-convert-an-array-to-object-in-php

Good to take heed over the warning of floating point and integer conversions on some systems.

Reply:
also if i just pass an array into the citeproc, it goes :boom: because it expects objects when it tries to do `clone`

Reply:
I have used this approach before with good success, fwiw.  You'd expect the casting `(Object)` style to be more robust, but nope.